### PR TITLE
[Bugfix][Model] Fix Qwen2

### PR DESCRIPTION
**Pull request by esmeetu**
Fix broken qwen2.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by ericzhou571** [comment]
Thank you for your contribution and for addressing this issue so effectively! I wanted to express my gratitude for this fix and highlight its importance to me and potentially others facing the same problem. Your work has directly contributed to the stability and functionality of our projects, and I am eagerly looking forward to seeing this pull request merged, especially before the v0.3.4 release.

I recently encountered an error after building from the main branch (dated 2024.03.21, commit: https://github.com/vllm-project/vllm/commit/865732342b4e3b8a4ef38f28a2a5bdb87cf3f970 ), which was causing deployment issues. The error message was as follows:

```
vllm/model_executor/models/qwen2.py", line 374, in load_weights
    param = params_dict[name]
KeyError: 'lm_head.weight'
```

This pull request, which includes changes to `vllm/model_executor/models/qwen2.py`, has successfully resolved the issue for me. After integrating these changes, the deployment proceeded smoothly without any errors. Furthermore, I've conducted tests on my own domain test set and found no discrepancies or issues resulting from this update.

File: tests/models/test_models.py [hunk 1]
```diff
@@ -5,21 +5,12 @@
...
-    "microsoft/phi-2",
-    "stabilityai/stablelm-3b-4e1t",
-    "allenai/OLMo-1B",
-    "bigcode/starcoder2-3b",
+    "facebook/opt-125m", "meta-llama/Llama-2-7b-hf",
+    "mistralai/Mistral-7B-v0.1", "Deci/DeciLM-7b", "tiiuae/falcon-7b", "gpt2",
+    "bigcode/tiny_starcoder_py", "EleutherAI/gpt-j-6b",
+    "EleutherAI/pythia-70m", "bigscience/bloom-560m", "mosaicml/mpt-7b",
+    "microsoft/phi-2", "stabilityai/stablelm-3b-4e1t", "allenai/OLMo-1B",
+    "bigcode/starcoder2-3b", "Qwen/Qwen1.5-0.5B"
```
Comment (WoosukKwon) [inline tests/models/test_models.py:13]: nit: Could you append `,` so that the linter does not squash the list?

Comment (esmeetu) [reply tests/models/test_models.py]: @WoosukKwon done

File: vllm/model_executor/models/qwen2.py [hunk 2]
```diff
@@ -349,7 +349,7 @@ def load_weights(self,
             ("gate_up_proj", "gate_proj", 0),
             ("gate_up_proj", "up_proj", 1),
         ]
-        params_dict = dict(self.named_parameters())
+        params_dict = dict(self.named_parameters(remove_duplicate=False))
```
Comment (WoosukKwon) [inline vllm/model_executor/models/qwen2.py:352]: Could you explain why this fixes the bug? Sorry I don’t have context for this.

Comment (esmeetu) [reply vllm/model_executor/models/qwen2.py:352]:
```
        if config.tie_word_embeddings:
            self.lm_head_weight = self.model.embed_tokens.weight
        else:
            self.lm_head = ParallelLMHead(config.vocab_size,
                                          config.hidden_size)
            self.lm_head_weight = self.lm_head.weight
```
Both have two same parameters which will be removed in `self.named_parameters(remove_duplicate=True)`.
So If we use another temp parameter `self.lm_head_weight`, we need use `remove_duplicate=False` like other models.

Comment (WoosukKwon) [reply vllm/model_executor/models/qwen2.py:352]: I see. Thanks! TIL

**Review by WoosukKwon — APPROVED** [review]
Lgtm thx!

**Comment by fyabc** [comment]
Hi @WoosukKwon and @esmeetu, it seems that my previous PR #3344 on Qwen2 is not fully merged into main branch (check not passed), it merges an incorrect old commit (with extra `self.lm_head_weight`).
In the latest version of #3344, I remove the temp parameter `self.lm_head_weight`, so I think we don't need to set `remove_duplicate=False`.
I create a new PR #3565 to fix it, please kindly check and give comments.

**Pull request merged**